src/inference/hmc_common.jl

```diff
@@ -0,0 +1,52 @@
+function sample_momenta(n::Int)
+  Float64[random(normal, 0, 1) for _=1:n]
+end
+
+function assess_momenta(momenta)
+  logprob = 0.
+  for val in momenta
+      logprob += logpdf(normal, val, 0, 1)
+  end
+  logprob
+end
+
+function add_choicemaps(a::ChoiceMap, b::ChoiceMap)
+  out = choicemap()
+
+  for (name, val) in get_values_shallow(a)
+    out[name] = val + b[name]
+  end
+
+  for (name, submap) in get_submaps_shallow(a)
+    out.internal_nodes[name] = add_choicemaps(submap, get_submap(b, name))
+  end
+
+  return out
+end
+
+function scale_choicemap(a::ChoiceMap, scale)
+  out = choicemap()
+
+  for (name, val) in get_values_shallow(a)
+    out[name] = val * scale
+  end
+
+  for (name, submap) in get_submaps_shallow(a)
+    out.internal_nodes[name] = scale_choicemap(submap, scale)
+  end
+
+  return out
+end
+
+function assess_momenta_trie(momenta_trie)
+  logprob = 0.
+  for (_, val) in get_values_shallow(momenta_trie)
+    logprob += logpdf(normal, val, 0, 1)
+  end
+
+  for (_, submap) in get_submaps_shallow(momenta_trie)
+    logprob += assess_momenta_trie(submap)
+  end
+
+  return logprob
+end
```

Comment:
Depending on usage below, it might be useful to optimize this operation by assuming you pass in a `Vector` -- a flattened version of `momenta_trie` to avoid the lookup overhead of recursively following references. That way, the logprob could be computed by evaluating the logprob of a single isotropic MvNormal.

Reply:
Makes sense